File: WinPaletter/GUI/Elements/ContainersAndBoxes/AlertBox.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D
Imports System.Drawing.Text

Namespace UI.WP

    <Description("AlertBox for WinPaletter UI")> Public Class AlertBox : Inherits ContainerControl

        Private borderColor, innerColor, textColor As Color

        Sub New()
            TabStop = False
            SetStyle(ControlStyles.AllPaintingInWmPaint Or ControlStyles.UserPaint Or ControlStyles.ResizeRedraw, True)
            DoubleBuffered = True
            Font = New Font("Segoe UI", 9)
            Size = New Size(200, 40)
            CenterText = False
            CustomColor = Color.FromArgb(0, 81, 210)
        End Sub

#Region "Properties"
        Enum Style
            Adaptive
            Simple
            Success
            Notice
            Warning
            Information
            Indigo
            Custom
        End Enum

        Enum Close
            Yes
            No
        End Enum

        Public Property AlertStyle As Style
            Get
                Return _alertStyle
            End Get
            Set(value As Style)
                _alertStyle = value
                Invalidate()
            End Set
        End Property

        Private _alertStyle As Style

        Public Property Image As Image
        Public Property CustomColor As Color
        Public Property CenterText As Boolean = False

        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        Public Overrides Property Text As String
#End Region

        Protected Overrides Sub OnPaint(e As System.Windows.Forms.PaintEventArgs)
            MyBase.OnPaint(e)
            Dim G As Graphics = e.Graphics
            G.SmoothingMode = SmoothingMode.AntiAlias
            G.TextRenderingHint = TextRenderingHint.SystemDefault
            DoubleBuffered = True
            Dim RTL As Boolean = (RightToLeft = 1)
            Dim DM As Boolean = My.Style.DarkMode

            Select Case _alertStyle
                Case Style.Simple
                    If DM Then
                        borderColor = Color.FromArgb(60, 60, 60)
                        innerColor = Color.FromArgb(50, 50, 50)
                        textColor = Color.FromArgb(150, 150, 150)
                    Else
                        borderColor = Color.FromArgb(190, 190, 190)
                        innerColor = Color.FromArgb(150, 150, 150)
                        textColor = Color.FromArgb(250, 250, 250)
                    End If

                Case Style.Success
                    If DM Then
                        borderColor = Color.FromArgb(60, 98, 79)
                        innerColor = Color.FromArgb(60, 85, 79)
                        textColor = Color.FromArgb(35, 169, 110)
                    Else
                        borderColor = Color.FromArgb(160, 198, 179)
                        innerColor = Color.FromArgb(140, 170, 155)
                        textColor = Color.FromArgb(135, 255, 210)
                    End If

                Case Style.Notice
                    If DM Then
                        borderColor = Color.FromArgb(70, 91, 107)
                        innerColor = Color.FromArgb(70, 91, 94)
                        textColor = Color.FromArgb(97, 185, 186)
                    Else
                        borderColor = Color.FromArgb(170, 191, 207)
                        innerColor = Color.FromArgb(130, 155, 155)
                        textColor = Color.FromArgb(180, 255, 255)
                    End If

                Case Style.Warning
                    If DM Then
                        borderColor = Color.FromArgb(202, 41, 56)
                        innerColor = Color.FromArgb(125, 20, 30)
                        textColor = Color.FromArgb(254, 142, 122)
                    Else
                        borderColor = Color.FromArgb(200, 171, 171)
                        innerColor = Color.FromArgb(150, 75, 75)
                        textColor = Color.FromArgb(255, 175, 175)
                    End If

                Case Style.Information
                    If DM Then
                        borderColor = Color.FromArgb(133, 133, 71)
                        innerColor = Color.FromArgb(120, 120, 71)
                        textColor = Color.FromArgb(254, 224, 122)
                    Else
                        borderColor = Color.FromArgb(233, 233, 171)
                        innerColor = Color.FromArgb(195, 195, 150)
                        textColor = Color.FromArgb(250, 250, 150)
                    End If


                Case Style.Indigo
                    If DM Then
                        borderColor = Color.FromArgb(65, 0, 170)
                        innerColor = Color.FromArgb(60, 0, 140)
                        textColor = Color.FromArgb(140, 0, 255).CB(0.35)
                    Else
                        borderColor = Color.FromArgb(165, 0, 225)
                        innerColor = Color.FromArgb(129, 0, 200)
                        textColor = Color.FromArgb(210, 110, 255)
                    End If

                Case Style.Custom

                    If DM Then
                        borderColor = CustomColor.CB(0.03)
                        innerColor = CustomColor.CB(0.01)
                        textColor = CustomColor.LightLight
                    Else
                        borderColor = CustomColor.CB(0.3)
                        innerColor = CustomColor.CB(0.1)
                        textColor = CustomColor.CB(0.7)
                    End If

                Case Style.Adaptive
                    If Image IsNot Nothing Then
                        Dim cc As Color = Image.AverageColor

                        If DM Then
                            borderColor = cc.Light(0.01)
                            innerColor = cc.Dark(0.05)
                            textColor = cc.LightLight
                        Else
                            borderColor = cc.Light(1)
                            innerColor = cc.LightLight.CB(0.35)
                            textColor = cc
                        End If

                    Else
                        If DM Then
                            borderColor = CustomColor.CB(0.03)
                            innerColor = CustomColor.CB(0.01)
                            textColor = CustomColor.LightLight
                        Else
                            borderColor = CustomColor.CB(0.3)
                            innerColor = CustomColor.CB(0.1)
                            textColor = CustomColor.CB(0.7)
                        End If
                    End If

            End Select

            G.Clear(GetParentColor)

            BackColor = innerColor

            Using br As New SolidBrush(innerColor) : G.FillRoundedRect(br, New Rectangle(0, 0, Width - 1, Height - 1)) : End Using
            Using P As New Pen(borderColor) : G.DrawRoundedRect_LikeW11(P, New Rectangle(0, 0, Width - 1, Height - 1)) : End Using

            Dim textY As Integer = CInt((Height - Text.Measure(Font).Height) / 2)
            Dim TextX As Integer = 5

            If Image IsNot Nothing Then G.DrawImage(Image, New Rectangle(If(Not RTL, 5, Width - 5 - Image.Width), 5, Image.Width, Image.Height))

            If Not CenterText Then
                If Image Is Nothing Then
                    Using br As New SolidBrush(textColor) : G.DrawString(Text, Font, br, New Rectangle(TextX, 0, Width, Height), ContentAlignment.MiddleLeft.ToStringFormat(RTL)) : End Using
                Else
                    If Not RTL Then
                        Using br As New SolidBrush(textColor) : G.DrawString(Text, Font, br, New Rectangle(10 + Image.Width, 7, Width - (5 + Image.Width), Height - 10), ContentAlignment.TopLeft.ToStringFormat) : End Using
                    Else
                        Using br As New SolidBrush(textColor) : G.DrawString(Text, Font, br, New Rectangle(0, 7, Width - (10 + Image.Width), Height - 10), ContentAlignment.TopLeft.ToStringFormat(RTL)) : End Using
                    End If
                End If
            Else
                Using br As New SolidBrush(textColor) : G.DrawString(Text, Font, br, New Rectangle(1, 0, Width, Height), ContentAlignment.MiddleCenter.ToStringFormat(RTL)) : End Using
            End If

        End Sub

    End Class

End Namespace

File: WinPaletter/GUI/Elements/ContainersAndBoxes/AlertBox.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Drawing.Drawing2D;
using System.Drawing.Text;
using System.Windows.Forms;
using System.Windows.Forms.Layout;

namespace WinPaletter.UI.WP
{
    [Description("AlertBox for WinPaletter UI")]
    public class AlertBox : Control
    {
        // Layout constants
        private const int PAD = 6;  // outer padding on every edge
        private const int PAD_AUTO = 2;  // small padding for AutoSize text-only mode
        private const int INNER_PAD = 4;  // gap between image and text
        private const int MIN_HEIGHT = 14; // never shrink below this
        private const int MIN_WIDTH = 20; // never shrink below this

        public AlertBox()
        {
            TabStop = false;
            SetStyle(
                ControlStyles.AllPaintingInWmPaint |
                ControlStyles.UserPaint |
                ControlStyles.ResizeRedraw |
                ControlStyles.SupportsTransparentBackColor,
                true);
            DoubleBuffered = true;
            Font = new Font("Segoe UI", 9f);
            Size = new Size(300, 40);
            CenterText = false;
            CustomColor = Color.FromArgb(0, 81, 210);
            BackColor = Color.Transparent;
        }

        #region Enums

        public enum Style
        {
            Adaptive,
            Simple,
            Success,
            Notice,
            Warning,
            Information,
            Indigo,
            Custom
        }

        #endregion

        #region Fields

        private Style _alertStyle;
        private Image _image;
        private bool _centerText;
        private string _text = string.Empty;
        private int _parentLevel;

        #endregion

        #region Properties

        [Category("Appearance")]
        public Style AlertStyle
        {
            get => _alertStyle;
            set
            {
                if (value == _alertStyle) return;
                _alertStyle = value;
                RefreshLayout();
            }
        }

        [Category("Appearance")]
        public Image Image
        {
            get => _image;
            set
            {
                _image = value;
                RefreshLayout();
            }
        }

        [Category("Appearance")]
        public Color CustomColor { get; set; }

        [Category("Appearance")]
        public bool CenterText
        {
            get => _centerText;
            set
            {
                if (value == _centerText) return;
                _centerText = value;
                RefreshLayout();
            }
        }

        /// <summary>
        /// When true the control resizes both width and height to exactly fit its content.
        /// Properly overrides the base Control.AutoSize so the designer serialises the value.
        /// </summary>
        [Browsable(true)]
        [Category("Layout")]
        [DefaultValue(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        public override bool AutoSize
        {
            get => base.AutoSize;
            set
            {
                if (base.AutoSize == value) return;
                base.AutoSize = value;
                // Tell the layout engine to re-evaluate immediately
                PerformLayout();
                RefreshLayout();
            }
        }

        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(MultilineStringEditor), typeof(UITypeEditor))]
        [Bindable(true)]
        public override string Text
        {
            get => _text;
            set
            {
                if (value == _text) return;
                _text = value ?? string.Empty;
                RefreshLayout();
            }
        }

        #endregion

        #region Layout helpers

        /// <summary>
        /// <summary>
        /// Returns the content rect, using zero padding when AutoSize and no image
        /// (the control is already sized to the raw text, so any inset would clip it).
        /// </summary>
        private Rectangle ContentRect(int controlWidth, int controlHeight)
        {
            int pad = (base.AutoSize && _image == null) ? PAD_AUTO : PAD;
            return new Rectangle(pad, pad,
                Math.Max(controlWidth - pad * 2, 0),
                Math.Max(controlHeight - pad * 2, 0));
        }

        private void CalcRects(Rectangle content, bool rtl, out Rectangle imageRect, out Rectangle textRect)
        {
            bool hasImage = _image != null;

            int imgW = hasImage ? _image.Width : 0;
            int imgH = hasImage ? _image.Height : 0;

            int textAvailW = content.Width - (hasImage ? imgW + INNER_PAD : 0);

            int imgX = content.Left;
            int textX = content.Left + (hasImage ? imgW + INNER_PAD : 0);

            if (rtl)
            {
                imgX = content.Right - imgW;
                textX = content.Left;
            }

            // Clamp all dimensions to zero — a negative width/height throws "Parameter is not valid"
            // when the control is very small (e.g. just dropped into the designer).
            imgW = Math.Max(imgW, 0);
            imgH = Math.Max(imgH, 0);
            int safeTextW = Math.Max(textAvailW, 0);
            int safeTextH = Math.Max(content.Height, 0);

            imageRect = hasImage
                ? new Rectangle(imgX, content.Top, imgW, imgH)
                : Rectangle.Empty;

            textRect = new Rectangle(textX, content.Top, safeTextW, safeTextH);
        }

        /// <summary>
        /// Measures text using the same GDI+ path as DrawString so AutoSize sizing
        /// and actual rendering are always consistent — no ellipsis on exact-fit strings.
        /// </summary>
        private SizeF MeasureTextExact(string text)
        {
            if (string.IsNullOrEmpty(text)) text = " ";
            using StringFormat sf = new StringFormat(StringFormatFlags.NoWrap | StringFormatFlags.NoClip);
            sf.Alignment = StringAlignment.Near;
            sf.LineAlignment = StringAlignment.Near;
            sf.Trimming = StringTrimming.None;
            // Use a Graphics from the screen at design-time, or from the control when live.
            using Graphics g = Graphics.FromHwnd(IntPtr.Zero);
            return g.MeasureString(text, Font, int.MaxValue, sf);
        }

        /// <summary>
        /// Calculates the preferred size to exactly fit image + text.
        /// Both axes are derived from content — the current Size is never used as a constraint.
        /// </summary>
        private Size CalcPreferredSize()
        {
            int imgW = _image?.Width ?? 0;
            int imgH = _image?.Height ?? 0;
            bool hasImage = _image != null;

            SizeF textSz = MeasureTextExact(_text);

            int contentW = (hasImage ? imgW + INNER_PAD : 0) + (int)Math.Ceiling(textSz.Width);
            int contentH = (int)Math.Ceiling(textSz.Height);
            if (hasImage) contentH = Math.Max(contentH, imgH);

            // Text-only AutoSize uses a small padding instead of zero or full PAD.
            int padW = hasImage ? PAD * 2 : PAD_AUTO * 2;
            int padH = hasImage ? PAD * 2 : PAD_AUTO * 2;

            return new Size(
                Math.Max(contentW + padW, MIN_WIDTH),
                Math.Max(contentH + padH, MIN_HEIGHT));
        }

        /// <summary>
        /// Called whenever any property that affects layout changes.
        /// Resizes the control when AutoSize is on, then invalidates.
        /// </summary>
        private void RefreshLayout()
        {
            if (base.AutoSize)
            {
                Size preferred = CalcPreferredSize();
                if (preferred != Size)
                    Size = preferred;
            }
            Invalidate();
        }

        // The designer and layout engine use it when AutoSize is true.
        public override Size GetPreferredSize(Size proposedSize) => base.AutoSize ? CalcPreferredSize() : base.GetPreferredSize(proposedSize);

        #endregion

        #region Overrides

        protected override void Dispose(bool disposing)
        {
            if (disposing) _image?.Dispose();
            base.Dispose(disposing);
        }

        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);
            _parentLevel = this.Level();
        }

        protected override void OnFontChanged(EventArgs e)
        {
            base.OnFontChanged(e);
            RefreshLayout();
        }

        protected override void OnResize(EventArgs e)
        {
            base.OnResize(e);
            // When something external resizes us while AutoSize is on, snap back.
            if (base.AutoSize) RefreshLayout();
        }

        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            // Intentionally empty — keeps the control background transparent.
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            // Must be first: redraws parent background behind this transparent control.
            InvokePaintBackground(this, e);

            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.AntiAlias;
            G.TextRenderingHint = DesignMode ? TextRenderingHint.ClearTypeGridFit : Program.Style.TextRenderingHint;

            bool rtl = (int)RightToLeft == 1;
            bool darkMode = Program.Style.DarkMode;

            Color borderColor, innerColor, textColor;

            Config.Scheme scheme1 = Enabled ? Program.Style.Schemes.Main : Program.Style.Schemes.Disabled;
            Config.Scheme scheme2 = Enabled ? Program.Style.Schemes.Secondary : Program.Style.Schemes.Disabled;
            Config.Scheme scheme3 = Enabled ? Program.Style.Schemes.Tertiary : Program.Style.Schemes.Disabled;

            switch (_alertStyle)
            {
                case Style.Simple:
                    borderColor = scheme1.Colors.Line_Hover(_parentLevel + 4);
                    innerColor = scheme1.Colors.Back(_parentLevel);
                    textColor = scheme1.Colors.ForeColor;
                    break;

                case Style.Success:
                    borderColor = scheme3.Colors.Line_Checked_Hover;
                    innerColor = scheme3.Colors.Back_Checked_Hover;
                    textColor = scheme3.Colors.ForeColor_Accent;
                    break;

                case Style.Notice:
                    using (var c = new Config.Colors_Collection(Color.FromArgb(115, 113, 6), default, darkMode))
                    {
                        borderColor = c.Line_Checked;
                        innerColor = c.Back_Checked;
                        textColor = c.ForeColor_Accent;
                    }
                    break;

                case Style.Warning:
                    borderColor = scheme2.Colors.Line_Checked;
                    innerColor = scheme2.Colors.Back_Checked;
                    textColor = scheme2.Colors.ForeColor_Accent;
                    break;

                case Style.Information:
                    borderColor = scheme1.Colors.Line_Checked;
                    innerColor = scheme1.Colors.Back_Checked;
                    textColor = scheme1.Colors.ForeColor_Accent;
                    break;

                case Style.Indigo:
                    using (var c = new Config.Colors_Collection(Color.FromArgb(76, 0, 142), default, darkMode))
                    {
                        borderColor = c.Line_Checked_Hover;
                        innerColor = c.Back_Checked_Hover;
                        textColor = c.ForeColor_Accent;
                    }
                    break;

                case Style.Custom:
                    using (var c = new Config.Colors_Collection(CustomColor, default, darkMode))
                    {
                        borderColor = c.Line_Checked_Hover;
                        innerColor = c.Back_Checked_Hover;
                        textColor = c.ForeColor_Accent;
                    }
                    break;

                case Style.Adaptive:
                default:
                    Color seed = _image?.AverageColor() ?? CustomColor;
                    using (var c = new Config.Colors_Collection(seed, default, darkMode))
                    {
                        borderColor = c.Line_Checked;
                        innerColor = c.Back_Checked;
                        textColor = c.ForeColor_Accent;
                    }
                    break;
            }

            if (textColor.A == 0)
                textColor = darkMode ? Color.White : Color.Black;

            // Background + border
            Rectangle outerRect = new(0, 0, Width - 1, Height - 1);
            using (SolidBrush br = new(innerColor)) G.FillRoundedRect(br, outerRect);
            using (Pen p = new(borderColor)) G.DrawRoundedRect(p, outerRect);

            // Content layout
            Rectangle content = ContentRect(Width, Height);
            CalcRects(content, rtl, out Rectangle imageRect, out Rectangle textRect);

            if (_image != null && imageRect != Rectangle.Empty) G.DrawImage(_image, imageRect);

            if (!string.IsNullOrEmpty(_text) && textRect.Width > 0 && textRect.Height > 0)
            {
                using StringFormat sf = new(StringFormatFlags.NoClip);
                sf.Alignment = _centerText || AutoSize ? StringAlignment.Center : (rtl ? StringAlignment.Far : StringAlignment.Near);
                sf.LineAlignment = _image != null ? StringAlignment.Near : StringAlignment.Center;
                sf.Trimming = StringTrimming.EllipsisCharacter;
                if (rtl) sf.FormatFlags |= StringFormatFlags.DirectionRightToLeft;

                using SolidBrush br = new(textColor);
                G.DrawString(_text, Font, br, textRect, sf);
            }

            base.OnPaint(e);
        }

        #endregion
    }
}